File: CevelopProject/bundles/com.cevelop.macronator/src/com/cevelop/macronator/checker/ProblemId.java
package com.cevelop.macronator.checker;

import ch.hsr.ifs.iltis.cpp.core.ast.checker.helper.IProblemId;


public enum ProblemId implements IProblemId<ProblemId> {

   //@formatter:off
   FUN_LIKE_MACRO("com.cevelop.macronator.plugin.ObsoleteFunctionLikeMacro"),
   UNUSED_MACRO("com.cevelop.macronator.plugin.UnusedMacro"),
   OBJECT_LIKE_MACRO("com.cevelop.macronator.plugin.ObsoleteObjectLikeMacro");
   //@formatter:on

    String id;

    ProblemId(String id) {
        this.id = id;
    }

    @Override
    public String getId() {
        return id;
    }

    public static ProblemId of(String string) {
        for (final ProblemId problemId : values()) {
            if (problemId.getId().equals(string)) {
                return problemId;
            }
        }
        throw new IllegalArgumentException("Illegal ProblemId: " + string);
    }

    @Override
    public ProblemId unstringify(String string) {
        return of(string);
    }

    @Override
    public String stringify() {
        return id;
    }

}

File: Plug-InsProject/bundles/com.cevelop.constificator/src/com/cevelop/constificator/checkers/ProblemId.kt
package com.cevelop.constificator.checkers

import java.util.Optional
import com.cevelop.constificator.core.util.type.Truelean
import ch.hsr.ifs.iltis.cpp.core.ast.checker.helper.IProblemId

enum class ProblemId private constructor(private val id: String) : IProblemId<ProblemId> {
	MISSING_QUALIFICATION(Markers.MissingQualification),
	POTENTIALLY_MISSING_QUALIFICATION(Markers.PotentiallyMissingQualification);

	override fun getId() = id
	
	override fun unstringify(string: String?): ProblemId? {
		return of(string)
	}

	override fun stringify(): String? {
		return id
	}

	companion object {
		fun of(string: String?): ProblemId? {
			for (problemId in values()) {
				if (problemId.id.equals(string)) {
					return problemId
				}
			}
			throw IllegalArgumentException("Illegal ProblemId: " + string!!)
		}

		fun of(decision: Truelean): Optional<ProblemId> {
			when (decision) {
				Truelean.YES -> return Optional.of(MISSING_QUALIFICATION)
				Truelean.MAYBE -> return Optional.of(POTENTIALLY_MISSING_QUALIFICATION)
				else -> return Optional.empty()
			}
		}
	}
}